File: yasmin/yasmin/state.py
# Copyright (C) 2023 Miguel Ángel González Santamarta
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <https://www.gnu.org/licenses/>.

from enum import Enum
from typing import Set
from threading import Lock
from abc import ABC, abstractmethod

import yasmin
from yasmin.blackboard import Blackboard


class StateStatus(Enum):
    """
    Enumeration representing the current status of a state.
    """

    IDLE = "idle"  # State is idle and ready to execute
    RUNNING = "running"  # State is currently executing
    CANCELED = "canceled"  # State execution has been canceled
    COMPLETED = "completed"  # State execution has completed successfully


class State(ABC):
    """
    Represents a state in a state machine.

    The State class defines a state that can execute actions and manage
    outcomes. It maintains information about its execution status
    and the possible outcomes of its execution.

    Attributes:
        _outcomes (Set[str]): The possible outcomes of this state.
        __status (StateStatus): Current status of the state.
        __status_lock (threading.Lock): Lock for thread-safe status operations.
    """

    def __init__(self, outcomes: Set[str]) -> None:
        """
        Constructs a State with a set of possible outcomes.

        Args:
            outcomes (Set[str]): A set of possible outcomes for this state.

        Raises:
            ValueError: If the provided outcomes set is empty.
        """

        ## The possible outcomes of this state.
        self._outcomes: Set = set()
        ## Current status of the state
        self.__status: StateStatus = StateStatus.IDLE
        ## Lock for thread-safe status operations
        self.__status_lock: Lock = Lock()

        if outcomes:
            self._outcomes.update(outcomes)
            self._outcomes: Set = sorted(self._outcomes)
        else:
            raise ValueError("A state must have at least one possible outcome.")

    def get_status(self) -> StateStatus:
        """
        Gets the current status of the State.

        Returns:
            StateStatus: The current status of the state.
        """
        with self.__status_lock:
            return self.__status

    def set_status(self, status: StateStatus) -> None:
        """
        Sets the status of the State.

        Args:
            status (StateStatus): The new status for the state.
        """
        with self.__status_lock:
            self.__status = status

    def is_idle(self) -> bool:
        """
        Checks if the State is idle.

        Returns:
            bool: True if the state is idle, otherwise False.
        """
        return self.get_status() == StateStatus.IDLE

    def is_running(self) -> bool:
        """
        Checks if the State is running.

        Returns:
            bool: True if the state is running, otherwise False.
        """
        return self.get_status() == StateStatus.RUNNING

    def is_canceled(self) -> bool:
        """
        Checks if the State is canceled.

        Returns:
            bool: True if the state is canceled, otherwise False.
        """
        return self.get_status() == StateStatus.CANCELED

    def is_completed(self) -> bool:
        """
        Checks if the State is completed.

        Returns:
            bool: True if the state is completed, otherwise False.
        """
        return self.get_status() == StateStatus.COMPLETED

    def __call__(self, blackboard: Blackboard = None) -> str:
        """
        Calls the state, transitioning from idle to running, executing, and then to completed.

        Args:
            blackboard (Blackboard, optional): An optional blackboard to share data with the state.

        Returns:
            str: The outcome of the state execution.
        """
        yasmin.YASMIN_LOG_DEBUG(f"Executing state '{self}'")

        self.set_status(StateStatus.RUNNING)

        if blackboard is None:
            blackboard = Blackboard()

        outcome = self.execute(blackboard)

        if outcome not in self._outcomes:
            self.set_status(StateStatus.IDLE)
            raise ValueError(
                f"Outcome '{outcome}' does not belong to the outcomes of the state '{self}'. "
                f"The possible outcomes are: {self._outcomes}"
            )

        # Mark as completed if not canceled
        if self.get_status() != StateStatus.CANCELED:
            self.set_status(StateStatus.COMPLETED)

        return outcome

    @abstractmethod
    def execute(self, blackboard: Blackboard) -> str:
        """
        Executes the state's specific logic.

        This method is intended to be overridden by derived classes to provide
        specific execution logic.

        Args:
            blackboard (Blackboard): A shared pointer to the Blackboard to use during
            execution.

        Returns:
            str: A string representing the outcome of the execution.
        """
        raise NotImplementedError("Subclasses must implement the execute method")

    def cancel_state(self) -> None:
        """
        Cancels the current state execution.

        This method sets the status to CANCELED and logs the action.
        """
        yasmin.YASMIN_LOG_INFO(f"Canceling state '{self}'")
        self.set_status(StateStatus.CANCELED)

    def get_outcomes(self) -> Set[str]:
        """
        Gets the set of possible outcomes for this state.

        Returns:
            Set[str]: A constant reference to the set of possible outcomes.
        """
        return self._outcomes

    def __str__(self) -> str:
        """
        Converts the state to a string representation.

        Returns:
            str: A string representation of the state.
        """
        return self.__class__.__name__

File: yasmin/src/yasmin/state.cpp
// Copyright (C) 2023 Miguel Ángel González Santamarta
//
// Licensed under the Apache License, Version 2.0 (the "License");
// you may not use this file except in compliance with the License.
// You may obtain a copy of the License at
//
//     http://www.apache.org/licenses/LICENSE-2.0
//
// Unless required by applicable law or agreed to in writing, software
// distributed under the License is distributed on an "AS IS" BASIS,
// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
// See the License for the specific language governing permissions and
// limitations under the License.

#include "yasmin/state.hpp"

#include <algorithm>
#include <exception>
#include <mutex>
#include <stdexcept>
#include <string>
#include <unordered_map>
#include <vector>

#include "yasmin/demangle.hpp"
#include "yasmin/logs.hpp"
#include "yasmin/state_utils.hpp"
#include "yasmin/types.hpp"

using namespace yasmin;

// ---------------------------------------------------------------------------
// External state storage — preserves ABI with plugins compiled against the
// older State class layout that did not include metadata or parameter storage.
// ---------------------------------------------------------------------------
namespace {
struct StateStorage {
  StateMetadata metadata;
  Blackboard::SharedPtr parameters = Blackboard::make_shared();
};

std::recursive_mutex &state_storage_mutex() {
  static std::recursive_mutex mtx;
  return mtx;
}

std::unordered_map<const State *, StateStorage> &state_storage_map() {
  static std::unordered_map<const State *, StateStorage> map;
  return map;
}
} // namespace

StateMetadata &State::get_metadata_ref() const {
  std::lock_guard<std::recursive_mutex> lock(state_storage_mutex());
  return state_storage_map()[this].metadata;
}

Blackboard::SharedPtr State::get_parameters_blackboard() const {
  std::lock_guard<std::recursive_mutex> lock(state_storage_mutex());
  return state_storage_map()[this].parameters;
}

State::State(Outcomes outcomes) {
  if (outcomes.empty()) {
    throw std::logic_error("A state must have at least one possible outcome.");
  }
  this->outcomes = std::move(outcomes);
}

State::~State() {
  std::lock_guard<std::recursive_mutex> lock(state_storage_mutex());
  state_storage_map().erase(this);
}

void State::set_status(StateStatus new_status) {
  this->status.store(new_status);
}

StateStatus State::get_status() const { return this->status.load(); }

bool State::is_idle() const noexcept {
  return this->get_status() == StateStatus::IDLE;
}

bool State::is_running() const noexcept {
  return this->get_status() == StateStatus::RUNNING;
}

bool State::is_canceled() const noexcept {
  return this->get_status() == StateStatus::CANCELED;
}

bool State::is_completed() const noexcept {
  return this->get_status() == StateStatus::COMPLETED;
}

std::string State::operator()(Blackboard::SharedPtr blackboard) {

  YASMIN_LOG_DEBUG("Executing state '%s'", this->to_string().c_str());
  this->set_status(StateStatus::RUNNING);

  // Inject default values for input keys that are missing from the blackboard
  const auto &input_keys = this->get_metadata_ref().input_keys;
  for (const auto &key_info : input_keys) {
    if (key_info.has_default && !blackboard->contains(key_info.name)) {
      YASMIN_LOG_DEBUG(
          "Injecting default value for input key '%s' in state '%s'",
          key_info.name.c_str(), this->to_string().c_str());
      key_info.inject_default(*blackboard, key_info.name);
    }
  }

  // Execute the specific logic of the state
  std::string outcome = this->execute(blackboard);

  // Check if the outcome is valid
  const auto &outcomes = this->get_outcomes();
  if (outcomes.find(outcome) == outcomes.end()) {

    // Mark as idle before throwing exception
    this->set_status(StateStatus::IDLE);

    if (outcome.empty()) {
      throw std::logic_error("State '" + this->to_string() +
                             "' returned an empty outcome. "
                             "Did you forget to override execute()?");
    }

    // Construct a string representation of the possible outcomes
    std::string outcomes_string =
        "[" +
        yasmin::join(outcomes, ", ", [](const std::string &o) { return o; }) +
        "]";

    // Throw an exception if the outcome is not valid
    throw std::logic_error("Outcome '" + outcome +
                           "' does not belong to the outcomes of "
                           "the state '" +
                           this->to_string() +
                           "'. The possible outcomes are: " + outcomes_string);
  }

  // Mark as completed if not canceled
  if (!this->is_canceled()) {
    this->set_status(StateStatus::COMPLETED);
  }

  return outcome; // Return the valid outcome
}

Outcomes const &State::get_outcomes() const noexcept { return this->outcomes; }

void State::set_description(const std::string &description) {
  this->get_metadata_ref().description = description;
}

const std::string &State::get_description() const {
  return this->get_metadata_ref().description;
}

void State::set_outcome_description(const std::string &outcome,
                                    const std::string &description) {
  if (this->outcomes.find(outcome) == this->outcomes.end()) {
    throw std::invalid_argument("Outcome '" + outcome +
                                "' is not part of state outcomes");
  }

  this->get_metadata_ref().outcome_descriptions[outcome] = description;
}

const std::string &
State::get_outcome_description(const std::string &outcome) const {
  if (this->outcomes.find(outcome) == this->outcomes.end()) {
    throw std::invalid_argument("Outcome '" + outcome +
                                "' is not part of state outcomes");
  }

  const auto &outcome_descriptions =
      this->get_metadata_ref().outcome_descriptions;
  const auto it = outcome_descriptions.find(outcome);

  static const std::string empty_description;
  if (it == outcome_descriptions.end()) {
    return empty_description;
  }

  return it->second;
}

const std::unordered_map<std::string, std::string> &
State::get_outcome_descriptions() const {
  return this->get_metadata_ref().outcome_descriptions;
}

void State::add_input_key(const BlackboardKeyInfo &key_info) {
  this->get_metadata_ref().input_keys.push_back(key_info);
}

void State::add_input_key(const std::string &key_name) {
  this->get_metadata_ref().input_keys.emplace_back(key_name);
}

void State::add_input_key(const std::string &key_name,
                          const std::string &description) {
  BlackboardKeyInfo info(key_name);
  info.description = description;
  this->get_metadata_ref().input_keys.push_back(info);
}

void State::add_output_key(const BlackboardKeyInfo &key_info) {
  BlackboardKeyInfo info = key_info;
  info.has_default = false;
  info.default_value.reset();
  info.default_value_type.clear();
  info.inject_default = {};
  this->get_metadata_ref().output_keys.push_back(info);
}

void State::add_output_key(const std::string &key_name) {
  this->get_metadata_ref().output_keys.emplace_back(key_name);
}

void State::add_output_key(const std::string &key_name,
                           const std::string &description) {
  BlackboardKeyInfo info(key_name);
  info.description = description;
  this->get_metadata_ref().output_keys.push_back(info);
}

const std::vector<BlackboardKeyInfo> &State::get_input_keys() const {
  return this->get_metadata_ref().input_keys;
}

const std::vector<BlackboardKeyInfo> &State::get_output_keys() const {
  return this->get_metadata_ref().output_keys;
}

void State::declare_parameter(const BlackboardKeyInfo &parameter_info) {
  auto &parameters = this->get_metadata_ref().parameters;

  auto it = std::find_if(parameters.begin(), parameters.end(),
                         [&](const BlackboardKeyInfo &info) {
                           return info.name == parameter_info.name;
                         });

  if (it == parameters.end()) {
    parameters.push_back(parameter_info);
  } else {
    *it = parameter_info;
  }

  if (parameter_info.has_default) {
    parameter_info.inject_default(*this->get_parameters_blackboard(),
                                  parameter_info.name);
  }
}

void State::declare_parameter(const std::string &parameter_name) {
  this->declare_parameter(BlackboardKeyInfo(parameter_name));
}

void State::declare_parameter(const std::string &parameter_name,
                              const std::string &description) {
  this->declare_parameter(BlackboardKeyInfo(parameter_name, description));
}

bool State::has_parameter(const std::string &parameter_name) const {
  return this->get_parameters_blackboard()->contains(parameter_name);
}

bool State::is_parameter_declared(const std::string &parameter_name) const {
  const auto &parameters = this->get_metadata_ref().parameters;
  return std::any_of(parameters.begin(), parameters.end(),
                     [&](const BlackboardKeyInfo &info) {
                       return info.name == parameter_name;
                     });
}

void State::copy_parameter_from(const State &source_state,
                                const std::string &source_parameter_name,
                                const std::string &target_parameter_name) {
  this->get_parameters_blackboard()->copy_value_from(
      *source_state.get_parameters_blackboard(), source_parameter_name,
      target_parameter_name);
}

const std::vector<BlackboardKeyInfo> &State::get_parameters() const {
  return this->get_metadata_ref().parameters;
}

const StateMetadata &State::get_metadata() const {
  return this->get_metadata_ref();
}

std::string State::to_string() const {
  return demangle_type(typeid(*this).name());
}
